ddtrace/tracer/stats.go

```diff
@@ -0,0 +1,331 @@
+	mu sync.Mutex
+
+	// buckets maintains a set of buckets, where the map key represents
+	// the starting point in time of that bucket, in nanoseconds.
+	buckets map[int64]*rawBucket
+
+	// stopped reports whether the concentrator is stopped (when non-zero)
+	stopped uint64
+
+	wg         sync.WaitGroup // waits for any active goroutines
+	bucketSize int64          // the size of a bucket in nanoseconds
+	stop       chan struct{}  // closing this channel triggers shutdown
+	cfg        *config        // tracer startup configuration
+}
+
+// newConcentrator creates a new concentrator using the given tracer
+// configuration c. It creates buckets of bucketSize nanoseconds duration.
+func newConcentrator(c *config, bucketSize int64) *concentrator {
+	return &concentrator{
+		In:         make(chan *aggregableSpan, 10000),
+		bucketSize: bucketSize,
+		stopped:    1,
+		buckets:    make(map[int64]*rawBucket),
+		cfg:        c,
+	}
+}
+
+// alignTs returns the provided timestamp truncated to the bucket size.
+// It gives us the start time of the time bucket in which such timestamp falls.
+func alignTs(ts, bucketSize int64) int64 { return ts - ts%bucketSize }
+
+// Start starts the concentrator. A started concentrator needs to be stopped
+// in order to gracefully shut down, using Stop.
+func (c *concentrator) Start() {
+	if atomic.SwapUint64(&c.stopped, 0) == 0 {
+		// already running
+		return
+	}
+	c.stop = make(chan struct{})
+	c.wg.Add(1)
+	go func() {
+		defer c.wg.Done()
+		tick := time.NewTicker(time.Duration(c.bucketSize) * time.Nanosecond)
+		defer tick.Stop()
+		c.runFlusher(tick.C)
+	}()
+	c.wg.Add(1)
+	go func() {
+		defer c.wg.Done()
+		c.runIngester()
+	}()
+}
+
+// runFlusher runs the flushing loop which sends stats to the underlying transport.
+func (c *concentrator) runFlusher(tick <-chan time.Time) {
+	for {
+		select {
+		case now := <-tick:
+			p := c.flush(now)
+			if len(p.Stats) == 0 {
+				// nothing to flush
+				continue
+			}
+			c.statsd().Incr("datadog.tracer.stats.flush_payloads", nil, 1)
+			c.statsd().Incr("datadog.tracer.stats.flush_buckets", nil, float64(len(p.Stats)))
+			if err := c.cfg.transport.sendStats(&p); err != nil {
+				c.statsd().Incr("datadog.tracer.stats.flush_errors", nil, 1)
+				log.Error("Error sending stats payload: %v", err)
+			}
+		case <-c.stop:
+			return
+		}
+	}
+}
+
+// statsd returns any tracer configured statsd client, or a no-op.
+func (c *concentrator) statsd() statsdClient {
+	if c.cfg.statsd == nil {
+		return &statsd.NoOpClient{}
+	}
+	return c.cfg.statsd
+}
+
+// runIngester runs the loop which accepts incoming data on the concentrator's In
+// channel.
+func (c *concentrator) runIngester() {
+	for {
+		select {
+		case ss := <-c.In:
+			c.statsd().Incr("datadog.tracer.stats.spans_in", nil, 1)
+			c.add(ss)
+		case <-c.stop:
+			return
+		}
+	}
+}
+
+// add adds ss into the concentrators internal stats buckets.
+func (c *concentrator) add(ss *aggregableSpan) {
```

Comment:
why ss and not s? What does ss stands for?

Reply:
Why not `span`?! (ducks)

Reply:
Ok ok guys 😄 The type was previously called `*spanSummary`, that's where the `ss` is coming from. Renaming to `s`.

> Why not `span`

That works, but it'll end up shadowing the `span` type, and that seems a bit confusing and error prone. I don't want to risk it.